src/main/java/com/google/api/gax/core/Page.java

```diff
@@ -32,27 +32,52 @@
 package com.google.api.gax.core;
 
 /**
- * Accessor for paged results from a list API method
+ * A Page object wraps an API list method response.
  *
- * <p>This is a subclass of Iterable where iterator() returns an Iterator object with the complete
- * listing result. If necessary it can perform more rpc calls to fetch more pages.
+ * <p>Callers can iterate over the Page object to get all elements returned in the page. The Page
+ * object also provides methods to retrieve additional pages using the page token, and to get the
+ * API request and response objects.
  */
-public interface PageAccessor<T> extends Iterable<T> {
+public interface Page<RequestT, ResponseT, ResourceT> extends Iterable<ResourceT> {
   /**
-   * Returns the values contained in this page.
-   * Note: This method is not thread-safe.
+   * Returns true if there are more pages that can be retrieved from the API.
    */
-  Iterable<T> getPageValues();
+  boolean hasNextPage();
 
   /**
-   * Returns the next page of results or {@code null} if no more results.
-   * Note: This method is not thread-safe.
+   * Returns the next page token from the response.
    */
-  PageAccessor<T> getNextPage();
+  Object getNextPageToken();
 
   /**
-   * Returns the token for the next page or {@code null} if no more results.
-   * Note: This method is not thread-safe.
+   * Retrieves the next Page object using the next page token.
    */
-  String getNextPageToken();
+  Page<RequestT, ResponseT, ResourceT> getNextPage();
```

Comment:
Probably update the comment to say what would happen if there is no more pages.

I am not sure but have you thought about just returning null instead of throwing an exception?